ansible_base/resource_registry/tasks/sync.py

```diff
@@ -0,0 +1,383 @@
+
+    def set_resource_local_variables():
+        """Inner caching function to avoid making unnecessary requests."""
+        nonlocal resource_data
+        nonlocal resource_type_name
+        nonlocal unavailable
+        if resource_data is None or resource_type_name is None:
+            resp = api_client.get_resource(manifest_item.ansible_id)
+            if 400 <= resp.status_code < 500:  # pragma: no cover
+                unavailable = True
+                return
+            resp.raise_for_status()
+            resource_data = resp.json()["resource_data"]
+            resource_type_name = resp.json()["resource_type"]
+
+    if local_managed_resource:
+        # Exists locally: Compare and Update
+        sc = local_managed_resource.content_type.resource_type.serializer_class
+        local_hash = sc(local_managed_resource.content_object).get_hash()
+        if manifest_item.resource_hash == local_hash:
+            return SyncResult(SyncStatus.NOOP, manifest_item)
+        set_resource_local_variables()
+        if unavailable:  # pragma: no cover
+            return SyncResult(SyncStatus.UNAVAILABLE, manifest_item)
+        # bind fetched resource_data for allowing reporting.
+        manifest_item.resource_data = resource_data
+        return _attempt_update_resource(
+            manifest_item,
+            local_managed_resource,
+            resource_data,
+        )
+    else:
+        # New: Create it locally
+        try:
+            set_resource_local_variables()
+            if unavailable:  # pragma: no cover
+                return SyncResult(SyncStatus.UNAVAILABLE, manifest_item)
+            manifest_item.resource_data = resource_data
+            resource_type = ResourceType.objects.get(name=resource_type_name)
+            Resource.create_resource(
+                resource_type=resource_type,
+                resource_data=resource_data,
+                ansible_id=manifest_item.ansible_id,
+                service_id=manifest_item.service_id,
+            )
+        except IntegrityError:
+            return SyncResult(SyncStatus.CONFLICT, manifest_item)
+        else:
+            return SyncResult(SyncStatus.CREATED, manifest_item)
+
+
+# https://docs.djangoproject.com/en/4.2/topics/async/#asgiref.sync.sync_to_async
+async_resource_sync = sync_to_async(resource_sync)
+
+
+@dataclass
+class SyncExecutor:
+    """Public Executor Implementing Sync and Async process."""
+
+    api_client: ResourceAPIClient = field(default_factory=create_api_client)
+    resource_type_names: list[ResourceTypeName] | None = None
+    retries: int = 0
+    retrysleep: int = 30
+    retain_seconds: int = 120
+    stdout: TextIOBase | None = None
+    unavailable: set = field(default_factory=set)
+    attempts: int = 0
+    deleted_count: int = 0
+    asyncio: bool = False
+
+    def write(self, text: str = ""):
+        """Write to assigned IO or simply ignores the text."""
+        self.stdout and self.stdout.write(text)
```

Comment:
more clear to write out `if self.stdout:`

Reply:
ok, I think the outcome is the same, I can do this change.